### PR TITLE
#10 - courses crud

### LINKED ISSUE #10: Courses management
We need:
- [ ] courses database structure (with internal id, abbreviation, full name, WYSIWYG description, semester number (not semester relation!) to chose and relations to field and form)
- [ ] CRUD actions

Current solution:

obraz

In current solution course has a field "is active" but I would move it to separate entity.

### PR DESCRIPTION
**Pull request by kamilpiech97**
Should close #10.

Added CRUD for courses.
Added tests.
Fixed toggling activity on semesters.

Index view:
image

Create view:
image

Edit view:
image

### CODE REVIEW

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: app/Enums/StudyForm.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        return [
+            "stationary" => __("stationary"),
+            "part-time" => __("part-time"),
+        ];
+    }
+
+    public static function abbreviationLabels(): array
+    {
+        return [
+            "part-time" => __("N"),
```
[the remark this answers is not in the document]
Comment (kamilpiech97) [reply app/Enums/StudyForm.php:23]: added

File: app/Http/Controllers/Dashboard/CourseController.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            "courses" => CourseResource::collection($courses),
+            "total" => Course::query()->count(),
+            "lastUpdate" => Course::query()->orderByDesc("updated_at")->first()?->updated_at->diffForHumans(),
+        ]);
+    }
+
+    public function create(): Response
+    {
+        return inertia("Dashboard/Course/Create", [
+            "semesters" => Semester::all(["id", "name"]),
```
[the remark this answers is not in the document]
Comment (kamilpiech97) [reply app/Http/Controllers/Dashboard/CourseController.php]: fixed

File: database/factories/CourseFactory.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+class CourseFactory extends Factory
+{
+    public function definition(): array
+    {
+        return [
+            "name" => fake()->asciify("******"),
+            "abbreviation" => fake()->asciify("*"),
+            "description" => fake()->text,
+            "semester" => Semester::factory()->create()->name,
```
[the remark this answers is not in the document]
Comment (kamilpiech97) [reply database/factories/CourseFactory.php:20]: fixed

File: database/migrations/2023_10_03_193906_create_courses_table.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    public function up(): void
+    {
+        Schema::create("courses", function (Blueprint $table): void {
+            $table->ulid("id")->primary();
+            $table->string("abbreviation");
+            $table->string("name");
+            $table->longText("description");
+            $table->string("semester");
+            $table->enum("type", ["lecture", "laboratory", "seminar", "workshop", "exercises", "project"]);
+            $table->enum("form", ["stationary", "part-time"]);
```
[the remark this answers is not in the document]
Comment (kamilpiech97) [reply database/migrations/2023_10_03_193906_create_courses_table.php]: fixed

File: app/Enums/ClassType.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Enums;
+
+enum ClassType: string
+{
+    case LABORATORY = "laboratory";
```
Comment (krzysztofrewak) [inline app/Enums/ClassType.php:9]: Why UPPERCASED?

File: app/Enums/Semester.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Enums;
+
+enum Semester: string
```
Comment (krzysztofrewak) [inline app/Enums/Semester.php:7]: I would drop this class and just use numeric semester numbers in the database.

Comment (krzysztofrewak) [reply app/Enums/Semester.php:7]: They could be romanized directly in model.

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: app/Enums/ClassType.php (hunk 5 above)
Comment (kamilpiech97) [reply app/Enums/ClassType.php:9]: I had uppercased enums in other projects too.

Comment (krzysztofrewak) [reply app/Enums/ClassType.php:9]: PHP documentation uses PascalCased cases. I would use that casing in all our projects.

File: app/Enums/Semester.php (hunk 6 above)
Comment (kamilpiech97) [reply app/Enums/Semester.php]: fixed

File: app/Enums/ClassType.php (hunk 5 above)
Comment (kamilpiech97) [reply app/Enums/ClassType.php:9]: I will stay with that

Comment (krzysztofrewak) [reply app/Enums/ClassType.php:9]: https://www.php-fig.org/per/coding-style/#9-enumerations

Comment (kamilpiech97) [reply app/Enums/ClassType.php]: changed to PascalCase

File: app/Enums/ClassType.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Enums;
+
+enum ClassType: string
+{
+    case Laboratory = "laboratory";
+    case LECTURE = "lecture";
```
Comment (krzysztofrewak) [inline app/Enums/ClassType.php:10]: :eyes:

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: app/Enums/ClassType.php (hunk 7 above)
Comment (kamilpiech97) [reply app/Enums/ClassType.php:10]: :roll_eyes:

Comment (kamilpiech97) [reply app/Enums/ClassType.php]: fixed already

**Review by krzysztofrewak — APPROVED** [review]

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Pull request merged**